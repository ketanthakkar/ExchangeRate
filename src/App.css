.main-container {
    display: flex;
    justify-content: center;
    margin-top: 20px;
}

.exchange-container {
    display: flex;
    align-items: center;
    flex-direction: column;
}

.form-container {
    margin-top: 20px;
}

.currency-container {
    width: 100%;
    display: flex;
}

.rate-container {
    text-align: center;
    color: #FFF;
    margin: 10px 0;
}
 
.rates {
    background: #fff;
    color: #0D7DBE;
    padding: 5px;
    border-radius: 35px;
    border: 1px solid #CCC;
}

select, input {
    height: auto;
    background: #fff;
    padding: 20px 5px;
    color: #2E4369;
    border: 0;
}

select {
    width: 30%;
    background-color: #476AAD;
    color: #FFF;
    font-size: 1.2rem;
    cursor: pointer;
}

input {
    width: 70%;
}

input[type=number]::-webkit-inner-spin-button, 
input[type=number]::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.amount-text {
    text-align: right;
    font-size: 1.5rem;
}