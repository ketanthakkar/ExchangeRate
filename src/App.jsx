import React from 'react';
import axios from 'axios';

import './App.css';

class App extends React.Component {

    constructor() {
        super();
        
        this.state = {
          baseCurrency: 'USD',
          targetCurrency: 'EUR',
          baseAmount: 0,
          targetAmount: 0,
          rates: [],
          currencies: [],
        };
    }

    componentDidMount() {
        this.callAPI(this.state.baseCurrency);
    }

    changeBaseCurrency = (e) => {
        this.setState({ baseCurrency: e.target.value});
        this.callAPI(e.target.value);
    }

    changeTargetCurrency = (e) => {
        this.setState({
            targetCurrency: e.target.value,
            targetAmount: Number.parseFloat(this.state.baseAmount * this.state.rates[e.target.value]).toFixed(2)});
    }
    
    changeBaseAmount = (e) => {
        this.setState({
        baseAmount: e.target.value,
        targetAmount: Number.parseFloat(e.target.value * this.state.rates[this.state.targetCurrency]).toFixed(2)
        });
    }

    changeTargetAmount = (e) => {
        this.setState({
            targetAmount: e.target.value,
            baseAmount: Number.parseFloat(e.target.value / this.state.rates[this.state.targetCurrency]).toFixed(2)
        });
    }

    validate = (e) => {
        let val = e.target.value;
        e.target.value = (val.indexOf(".") >= 0) ? (val.substr(0, val.indexOf(".")) + val.substr(val.indexOf("."), 3)) : val;
    }
  
  //API call to get latest data
    callAPI = (base) => {
        const url = `https://api.exchangeratesapi.io/latest?base=${base}`;
        
        return axios.get(url)
        .then(data => { 
            this.setState({
                rates: data.data['rates'],
                currencies: Object.keys(data.data['rates']).sort(),
                targetAmount: Number.parseFloat(this.state.baseAmount * data.data['rates'][this.state.targetCurrency]).toFixed(2)
            })
            this.startTimer();
        });
    }

    startTimer = () => {
        this.refreshRate = setTimeout(() => {
            this.callAPI(this.state.baseCurrency);
        }, 10000);
    }

    componentWillUnmount() {
        clearTimeout(this.refreshRate);
    }
    
    render() {
        const {currencies, rates, baseCurrency, baseAmount, targetAmount, targetCurrency} = this.state;
  
        const selectCurrency = currencies.map(currency =>
            <option key={currency} value={currency}> {currency} </option>      
          );

        return (
            <section className="main-container">
                <div className="exchange-container">
                    <h2>Exchange</h2>
                    <div className="form-container">
                        <form>
                            <div className="currency-container">               
                                <select name="baseCurrency" value={baseCurrency} onChange={this.changeBaseCurrency}>
                                    {selectCurrency}
                                    <option>{baseCurrency}</option>
                                </select>
                                <input className="amount-text" 
                                        type='number' 
                                        placeholder="0.00" 
                                        value={baseAmount} 
                                        onInput={this.validate} 
                                        onChange={this.changeBaseAmount}>
                                </input>
                            </div>
                            <div className="rate-container">
                                <span className="rates">
                                    1 {baseCurrency} = {rates && rates[targetCurrency]} {targetCurrency}
                                </span>
                            </div>
                            <div className="currency-container target-container">
                                <select name="targetCurrency" value={targetCurrency} onChange={this.changeTargetCurrency}>
                                    {selectCurrency}
                                </select> 
                                <input className="amount-text"
                                       type='number'
                                       placeholder="0.00"
                                       value={targetAmount}
                                       onInput={this.validate}
                                       onChange={this.changeTargetAmount}>
                                </input>
                            </div>                    
                        </form>
                    </div>
                </div>
            </section>
        );
    }
}

export default App;