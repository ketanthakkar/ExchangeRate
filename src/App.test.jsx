import React from 'react';
import { shallow, mount } from 'enzyme';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import App from './App'

describe('App Snapshot', () => {
    it('renders', () => {
      const component = shallow(<App />);
      expect(component).toMatchSnapshot();
    });
});

describe('App', () => {
    it('renders without crashing', () => {
        const wrapper = shallow(<App />);

        expect(wrapper).toBeTruthy();
    });

    it('should select correct base currency on change', () => {
        const wrapper = mount(<App />)
        wrapper.find('[name="baseCurrency"]').simulate('change', {target: {value: 'EUR'}});
        expect(wrapper.find('[name="baseCurrency"]').props().value).toBe('EUR');
    });

    it('should select correct target currency on change', () => {
        const wrapper = mount(<App />)
        wrapper.find('[name="targetCurrency"]').simulate('change', {target: {value: 'EUR'}});
        expect(wrapper.find('[name="targetCurrency"]').props().value).toBe('EUR');
    });

    it('should update state with response data', () => {
        const url = 'https://api.exchangeratesapi.io/latest?base=base';
        const mock = new MockAdapter(axios);
        const data = { response: true };
        mock
          .onGet(url)
          .reply(200, data);
    
    
        const component = shallow(<App />);
        const instance = component.instance();
    
        //instance.callAPI
        instance.callAPI('USD').then(response => {
          expect(response).toEqual(data);
          done();
        });
  
    });
});
