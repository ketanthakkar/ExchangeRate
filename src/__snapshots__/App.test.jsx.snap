// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshot renders 1`] = `
<section
  className="main-container"
>
  <div
    className="exchange-container"
  >
    <h2>
      Exchange
    </h2>
    <div
      className="form-container"
    >
      <form>
        <div
          className="currency-container"
        >
          <select
            name="baseCurrency"
            onChange={[Function]}
            value="USD"
          >
            <option>
              USD
            </option>
          </select>
          <input
            className="amount-text"
            onChange={[Function]}
            onInput={[Function]}
            placeholder="0.00"
            type="number"
            value={0}
          />
        </div>
        <div
          className="rate-container"
        >
          <span
            className="rates"
          >
            1 
            USD
             = 
             
            EUR
          </span>
        </div>
        <div
          className="currency-container target-container"
        >
          <select
            name="targetCurrency"
            onChange={[Function]}
            value="EUR"
          />
          <input
            className="amount-text"
            onChange={[Function]}
            onInput={[Function]}
            placeholder="0.00"
            type="number"
            value={0}
          />
        </div>
      </form>
    </div>
  </div>
</section>
`;
